src/cache_cf.cc

```diff
@@ -633,6 +634,75 @@ parseConfigFile(const char *file_name)
     }
 }
 
+/*
+ * The templated functions below are essentially ConfigParser methods. They are
+ * not implemented as such because our generated code calling them is the only
+ * code that can instantiate implementations for each T -- we cannot place these
+ * definitions into ConfigParser.cc unless cf_parser.cci is moved there.
+ */
```

Comment:
Since these are static templates they should be added to src/config/gadgets.h and included by cf_parser.cci

Reply:
I cannot find a logical connection between "these are static templates" and "they should be added to src/config/gadgets.h and included by cf_parser.cci". If anything, the opposite (i.e. "should _not_ be ...") statement would be logically sound. Unfortunately, I do not know what problem you are trying to fix here. Moreover, I see no in-scope problems worth fixing here. Thus, it is difficult for me to suggest specific fixes to this change request that we may be able to agree on.

These templates should not be added to any C++ header file because they are not shared across translation units -- only one translation unit needs them.

Moreover, it is conceptually wrong to create header files just to be included in a _generated_ .cci file (which is itself included into a .cc file). We should keep cf_parser.cci as small/simple as possible, containing code that has to be generated and virtually nothing else. Everything that should not be generated goes into cache_cf.cc which includes cf_parser.cci (or into other manually-written source files). Our cf_parser.cci does not contain `#include` statements, and we should keep it that way (until some `#include` statements must be generated).

Reply:
Being static templates they can be added to any .h file.

In context of my request to create src/config/ library for this polished API, they should have a .h there away from the old legacy cruft.

Under our current coding style either each one gets a file named after itself, or being too complex for forward.h they go into gadgets.h. If you can think of a filename descriptive of the group they can use that. I am not fixed on any particular name.

Since only the generated code in .cci uses these templates the best place to include is there. I will accept the cache_cf.cc if you insist though.

Reply:
> Being static templates they can be added to any .h file.

Many things can be added to any .h file. Most such things should not be. This is not an argument that can support or justify placing something into a header file!

> In context of my request to create src/config/ library for this polished API, they should have a .h there away from the old legacy cruft.

These functions should not be in a header file, as I tried to explain in my previous [response](https://github.com/squid-cache/squid/pull/811#discussion_r625812345). I have no new arguments, but if you need me to clarify the existing ones (which your response completely ignored!), please ask questions or point to flaws in my logic.

Yes, it is nice to separate new code from the "old cruft" on a source file basis, but _not_ at the expense of lowering the quality of that new code. FWIW, most of this new code can still be easily detected in this case by searching for Configuration::Component.

> Under our current coding style either each one gets a file named after itself

I see no coding style rules that require these local functions to be in a header file (or even a .cc file named after them). If there are such rules, they should probably adjusted/clarified!

> Since only the generated code in .cci uses these templates the best place to include is there.

Virtually any manually written code that can be placed outside of generated code should be placed outside, for many reasons (please let me know if I should detail them). Whether the generated code is the only user of that manually written code has no bearing on this rule of thumb.

> I will accept the cache_cf.cc if you insist though.

Yes, I do "insist" that placing these functions in a header in this PR would be a mistake. I do so not because I can insist on arbitrary things, but because I have outlined solid reasons to support my position and have been given no reasons that would justify the move (except, perhaps, the "place new code into new files" principle that I worked hard to extract and that cannot be satisfied without violating more important principles here).